Minor optimizations to SerializedScriptValue

cdumez opened the pull request:
#### 473600ff56268ca69413333c0d7ccc611ea5c6d0
<pre>
Minor optimizations to SerializedScriptValue
<a href="https://bugs.webkit.org/show_bug.cgi?id=259794">https://bugs.webkit.org/show_bug.cgi?id=259794</a>

Reviewed by Ryosuke Niwa and Darin Adler.

Minor optimizations to SerializedScriptValue:
1. Encode boolean values as uint8_t instead of int32_t to save space.
2. Move String in when constructing a CachedString
3. Avoid double HashMap lookup and code duplication in fillTransferMap()

* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::fillTransferMap):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneSerializer::write):
(WebCore::CloneDeserializer::CachedString::CachedString):
(WebCore::CloneDeserializer::read):
(WebCore::CloneDeserializer::readStringData):

Canonical link: <a href="https://commits.webkit.org/266612@main">https://commits.webkit.org/266612@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/10d3c595e14d73ac3066b0a03cb664542dfdb00c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14250 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14564 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14904 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15986 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13495 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14355 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17077 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14642 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16168 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14427 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14990 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12090 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16707 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12274 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12851 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19862 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13352 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13015 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16217 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13560 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11417 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12843 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/12711 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3456 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17180 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13405 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -913,29 +914,13 @@ class CloneSerializer : CloneBase {
             JSC::JSValue value = toJS(m_lexicalGlobalObject, globalObject, input[i].get());
-            JSC::JSObject* obj = value.getObject();
-            if (obj && !result.contains(obj))
+            if (auto* obj = value.getObject())
```
I would have written:
```
    if (auto* object = toJS(globalObject, globalObject, input[i].get()).getObject())
```
Different variable names, fewer locals, not passing the same object pointer two different ways.

darinadler commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1624,7 +1609,6 @@ class CloneSerializer : CloneBase {
                     return true;
                 } else if (m_context == SerializationContext::CloneAcrossWorlds) {
-                    // FIXME: Only in clone accross worlds mode.
                     write (InMemoryMessagePortTag);
```
Strange space here after `write`.

darinadler commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1878,7 +1862,7 @@ class CloneSerializer : CloneBase {
     void write(bool b)
     {
-        writeLittleEndian(m_buffer, static_cast<int32_t>(b));
+        writeLittleEndian(m_buffer, static_cast<uint8_t>(b));
```
Could have written this:
```
    write(static_cast<uint8_t>(b));
```
Not sure that is better.

darinadler commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -2858,10 +2839,17 @@ class CloneDeserializer : CloneBase {
-        if (!readLittleEndian(integer) || integer > 1)
-            return false;
-        b = !!integer;
+        if (m_version >= 14) {
```
I don’t understand why we need this check. Does cloning actually need compatibility with older versions?

cdumez replied:
As far as I know, we need to preserve backward compatibility. I think it is because such serialization can be stored via IndexedDB. This is the reason we have versioning.

darinadler commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -4064,7 +4052,7 @@ class CloneDeserializer : CloneBase {
     JSValue readBigInt()
     {
-        uint8_t sign = 0;
+        bool sign;
```
Seems like this would not work if we are reading from a `m_version < 14` buffer. So if we do need the check in the other function, we probably need something different here.

cdumez replied:
Ohhh, you're right. This was always encoded as a uint8_t so my read(bool&) will fail if `m_version < 14`. I'll address.

darinadler commented on LayoutTests/fast/storage/serialized-script-value.html:
```diff
@@ -6,7 +6,7 @@
         <script>
 
-const currentVersion = 0x0d;
+const currentVersion = 0x0e;
```
Why is this in hex? If this needs to be kept in sync with the one in the `.cpp` file I would think we’d write 14 in both places.

darinadler commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -2858,10 +2839,17 @@ class CloneDeserializer : CloneBase {
-        b = !!integer;
+        if (m_version >= 14) {
+            uint8_t integer;
+            if (!readLittleEndian(integer) || integer > 1)
```
We could just write `read` here instead of `readLittleEndian`.

darinadler commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -2858,10 +2839,17 @@ class CloneDeserializer : CloneBase {
+            b = !!integer;
+        } else {
+            int32_t integer;
+            if (!readLittleEndian(integer) || integer > 1)
```
We could just write `read` here instead of `readLittleEndian`.

cdumez committed:
Minor optimizations to SerializedScriptValue
https://bugs.webkit.org/show_bug.cgi?id=259794

Reviewed by Ryosuke Niwa and Darin Adler.

Minor optimizations to SerializedScriptValue:
1. Encode boolean values as uint8_t instead of int32_t to save space.
2. Move String in when constructing a CachedString
3. Avoid double HashMap lookup and code duplication in fillTransferMap()

* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::fillTransferMap):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneSerializer::write):
(WebCore::CloneDeserializer::CachedString::CachedString):
(WebCore::CloneDeserializer::read):
(WebCore::CloneDeserializer::readStringData):

Canonical link: https://commits.webkit.org/266612@main
